util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,339 @@
+{
+	if message_types.get(message_type).is_none() {
+		return None;
+	}
+
+	let mut types = LinkedHashSet::new();
+	types.insert(message_type);
+	let mut deps = LinkedHashSet::new();
+
+	while let Some(item) = types.pop_back() {
+		if let Some(fields) = message_types.get(item) {
+			deps.insert(item);
+
+			for field in fields {
+				// seen this type before? or not a custom type skip
+				if deps.contains(&*field.type_) || !message_types.contains_key(&*field.type_) {
+					continue;
+				}
+				types.insert(&*field.type_);
+			}
+		}
+	};
+
+	return Some(deps)
+}
+
+fn encode_type(message_type: &str, message_types: &MessageTypes) -> Result<String> {
+	let deps = {
+		let mut temp = build_dependencies(message_type, message_types).ok_or_else(|| ErrorKind::NonExistentType)?;
+		temp.remove(message_type);
+		let mut temp = temp.into_iter().collect::<Vec<_>>();
+		(&mut temp[..]).sort_unstable();
+		temp.insert(0, message_type);
+		temp
+	};
+
+	let encoded = deps
+		.into_iter()
+		.filter_map(|dep| {
+			message_types.get(dep).map(|field_types| {
+				let types = field_types
+					.iter()
+					.map(|value| format!("{} {}", value.type_, value.name))
+					.join(",");
+				return format!("{}({})", dep, types);
+			})
+		})
+		.collect::<Vec<_>>()
+		.concat();
+	Ok(encoded)
+}
+
+fn type_hash(message_type: &str, typed_data: &MessageTypes) -> Result<H256> {
+	Ok(keccak(encode_type(message_type, typed_data)?))
+}
+
+fn encode_data(
+	parser: &Parser,
+	message_type: &Type,
+	message_types: &MessageTypes,
+	value: &Value,
+	field_name: Option<&str>
+) -> Result<Vec<u8>>
+{
+	let encoded = match message_type {
+		Type::Array(array_type) => {
+			let mut items = vec![];
+
+			for item in value.as_array().ok_or_else(|| serde_error("array", field_name))? {
+				let mut encoded = encode_data(parser, &*array_type, &message_types, item, field_name)?;
+				items.append(&mut encoded);
+			}
+
+			keccak(items).0.to_vec()
+		}
+
+		Type::Custom(ref ident) if message_types.get(&*ident).is_some() => {
+			let type_hash = (&type_hash(ident, &message_types)?).to_vec();
+			let mut tokens = encode(&[EthAbiToken::FixedBytes(type_hash)]);
+
+			for field in message_types.get(ident).expect("Already checked in match guard; qed") {
+				let value = &value[&field.name];
+				let type_ = parser.parse_type(&*field.type_)?;
+				let mut encoded = encode_data(parser, &type_, &message_types, &value, Some(&*field.name))?;
+				tokens.append(&mut encoded);
+			}
+
+			keccak(tokens).0.to_vec()
+		}
+
+		Type::Bytes(size) => {
+			let string = value.as_str().ok_or_else(|| serde_error("string", field_name))?;
+			if string.len() < 2 {
+				return Err(ErrorKind::HexParseError(
+					format!("Expected a 0x-prefixed string of even length, found {} length string", string.len()))
+				)?
+			}
+			let mut bytes = H256::from_str(&string[2..]).map_err(|err| ErrorKind::HexParseError(format!("{}", err)))?;
+			if *size == 32 {
```

Comment:
I don't understand what's going on here, we hash only in case the size is 32? Can you point to a part of the spec defining that?

Reply:
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-typed-structured-data-%F0%9D%95%8A

Reply:
@maciejhirsz looks like lunarity parses `bytes` as `Bytes32`

Reply:
Don't see anyhing in the link that would suggest that we hash in case the size is 32 bytes. It actually says somehing completely different:

> bytes1 to bytes31 are arrays with a beginning (index 0) and an end (index length - 1), they are zero-padded at the end to bytes32 and encoded in beginning to end order. This corresponds to their encoding in ABI v1 and v2.
> The dynamic values bytes and string are encoded as a keccak256 hash of their contents.

So we should only hash if bytes vector is (strictly) greater than 32 bytes and just pad the results in case it's smaller.

Reply:
> looks like lunarity parses bytes as Bytes32

@tomusdrw yeah i made the incorrect assumption that `bytes32` == `bytes`